pkg/jx/cmd/delete_app.go

```diff
@@ -90,23 +93,82 @@ func NewCmdDeleteApp(f Factory, in terminal.FileReader, out terminal.FileWriter,
 	cmd.Flags().StringVarP(&options.SelectFilter, "filter", "f", "", "Filter the list of apps to those containing this text")
 	cmd.Flags().StringVarP(&options.Timeout, optionTimeout, "t", "1h", "The timeout to wait for the promotion to succeed in the underlying Environment. The command fails if the timeout is exceeded or the promotion does not complete")
 	cmd.Flags().StringVarP(&options.PullRequestPollTime, optionPullRequestPollTime, "", "20s", "Poll time when waiting for a Pull Request to merge")
+	// TODO - Create an Application CRD that gets populated with the org when an application is created/imported to store this.
+	cmd.Flags().StringVarP(&options.Org, "org", "o", "", "github organisation/project name that source code resides in. Temporary workaround until the platform can determine this automatically")
 	cmd.Flags().BoolVarP(&options.BatchMode, "batch-mode", "b", false, "Run without being prompted. WARNING! You will not be asked to confirm deletions if you use this flag.")
 
 	return cmd
 }
 
 // Run implements this command
 func (o *DeleteAppOptions) Run() error {
+	err := o.init()
+	if err != nil {
+		return err
+	}
+
+	isProw, err := o.isProw()
+	if err != nil {
+		return err
+	}
+
+	var deletedApplications []string
+	if isProw {
+		deletedApplications, err = o.deleteProwApp()
+	} else {
+		deletedApplications, err = o.deleteJenkinsApp()
+	}
+
+	if err != nil {
+		return errors.Wrapf(err, "deleting application")
+	}
+	log.Infof("Deleted Application(s): %s\n", util.ColorInfo(strings.Join(deletedApplications, ",")))
+	return nil
+}
+
+func (o *DeleteAppOptions) deleteProwApp() (deletedApplications []string, err error) {
+	if o.Org == "" {
+		return deletedApplications, errors.New("--org must be supplied")
+	}
+	envMap, _, err := kube.GetOrderedEnvironments(o.jxClient, "")
+	currentUser, err := user.Current()
+	if err != nil {
+		return deletedApplications, errors.Wrap(err, "getting current user")
+	}
+
+	kubeClient, ns, err := o.Factory.CreateKubeClient()
+	if err != nil {
+		return deletedApplications, errors.Wrap(err, "getting kube client")
+	}
+
+	for _, appName := range o.Args {
+		for _, env := range envMap {
+			err = o.deleteAppFromEnvironment(env, appName, currentUser.Username)
+			if err != nil {
+				return deletedApplications, errors.Wrapf(err, "deleting application %s from environment %s", appName, env.Name)
+			}
+		}
+		repo := []string{o.Org + "/" + appName}
+		err = prow.DeleteApplication(kubeClient, repo, ns)
+		if err != nil {
+			return deletedApplications, errors.Wrapf(err, "deleting prow config for %s", appName)
+		}
+		deletedApplications = append(deletedApplications, appName)
+	}
+	return
+}
+
+func (o *DeleteAppOptions) deleteJenkinsApp() (deletedApplications []string, err error) {
 	args := o.Args
 
 	jenk, err := o.JenkinsClient()
 	if err != nil {
-		return err
+		return
```

Comment:
bare returns are generally not used elswhere in the code, should we be introducing them here (there is an issue to remove the functionality from golang)

Reply:
I'll take a look

Reply:
It's not a bare return anymore, it's a named return (I think) - it returns `(deletedApplications []string, err error)`

Reply:
I think the naming has changed over the time (bare = naked = named).  I may be wrong on that though